Expanded the random number generator in the `platform_util.c` file

piotr-now opened the pull request:
## Description
The earlier implementation had two problems: the random generator always returned 0 if the MBEDTLS_ENTROPY_HARDWARE_ALT flag was not defined and there was no protection needed if the HW RNG was malfunctioning. Both these problems have been solved in this PR by adding the linear congruential generator algorithm as a non-cryptographic random number generator.

## Status
**READY**

AndrzejKurek commented on include/mbedtls/platform_util.h:
```diff
@@ -275,8 +275,8 @@ uint32_t mbedtls_platform_random_in_range( uint32_t num );
  *
- * \note        Currently the function is dependent of hardware providing an
- *              rng with MBEDTLS_ENTROPY_HARDWARE_ALT.
+ * \note        This function depends on whether the MBEDTLS_FI_COUNTERMEASURES
```
Please expand on how it depends on this flag. Will it stop working without it? Does it alter its behaviour?

piotr-now replied:
It works only I the MBEDTLS_FI_COUNTERMEASURES is defined. I added an explanation to the header file.

AndrzejKurek commented on library/platform_util.c:
```diff
@@ -63,7 +63,7 @@
  * random number generator will be changed. This is used only if the
  * MBEDTLS_ENTROPY_HARDWARE_ALT option is enabled. */
-#define RAND_SEED_LIFE  100
+#define RAND_SEED_LIFE  10000
```
Commit 0962e6b7574d57345b409796d90f0c3f725bcc17 message is too long.

piotr-now replied:
Actually, I moved this commit to the previous one

AndrzejKurek commented on library/platform_util.c:
```diff
@@ -51,9 +51,20 @@
+ * random number generator. */
+#define RAND_MULTIPLIER 1103515245
+#define RAND_INCREMENT  12345
+#define RAND_MODULUS    0x80000000
```
Why is this 0x80000000? How were the multiplier and incrementation values picked?

piotr-now replied:
All three parameters are standard values for a rand() function used in glibc and ANSI C

shelib01 replied:
@piotr-now , lets add comment and reference to rand() function of ANSI C

piotr-now replied:
Added

AndrzejKurek commented on library/platform_util.c:
```diff
@@ -172,50 +183,77 @@ int mbedtls_platform_memcmp( const void *buf1, const void *buf2, size_t num )
+ * 
+ * The value of the returned number is in the range [0; 0x80000000).
+ */
+static uint32_t mbedtls_platform_random( void )
```
Why is such range used? It causes a need to call `mbedtls_platform_random_uint32` twice and merge the results.

piotr-now replied:
This is due to the use of the modulo 0x80000000 value. This is a standard value used by glibc and ANSI C

AndrzejKurek commented on library/platform_util.c:
```diff
@@ -211,29 +212,24 @@ static uint32_t mbedtls_platform_random( void )
         if( olen == sizeof( hw_random ) )
         {
-            random_seed = hw_random % RAND_MODULUS;
+            random_state ^= hw_random;
```
Why was this changed to a XOR?

piotr-now replied:
This change was suggested by @armask to do entropy mixing instead of changing the seed.

piotr-now commented:
I did changes what you asked for, but it seems to me that until PR https://github.com/ARMmbed/mbedtls/pull/3514 is merged with the baremetal branch, I won't be able to resolve the conflicts. My changes require a few commits that are already in baremetal and in my PR, but which are not in https://github.com/ARMmbed/mbedtls/pull/3514. 

I will resolve conflicts when PR https://github.com/ARMmbed/mbedtls/pull/3514 is merged or rebased with the baremetal

piotr-now commented:
Force push after rebase on baremetal branch

shelib01 commented on library/platform_util.c:
```diff
@@ -172,51 +183,79 @@ int mbedtls_platform_memcmp( const void *buf1, const void *buf2, size_t num )
     }
+    else if( num <= 0xffff )
+    {
+        result = mbedtls_platform_random_uint16() % num;
```
@piotr-now ת why do we need mbedtls_platform_random_uint16 ? Why we cannot use always mbedtls_platform_random_uint32 ?

piotr-now replied:
Actually, we don't need mbedtls_platform_random_uint16(). This is only related to optimization because mbedtls_platform_random_in_range() is most often called for values below 0xffff and one call of mbedtls_platform_random_uint16() is enough. For values above 0xffff the mbedtls_platform_random_uint16() function must be called twice and this is done by mbedtls_platform_random_uint32()

shelib01 replied:
But you always generate uint32_t in hw poll function.  What am i missing?

shelib01 replied:
Anyway, lets generate always 32. It will save as a little bit code size

piotr-now replied:
Done

shelib01 replied:
@piotr-now , But there is still random32 functions that calls to random16 twice. And random16 function returns 32bit result twice.  Do we have security reason for double random ?

piotr-now replied:
There are two reasons why I did this. The first results from the modulo value in the random generator. If the generator were to return the uint32_t value, the modulo value would have to be 2^32 (for the 32-bit arch the modulo value would not be at all). The second reason was more important to me and it is that the random generator would have to return the entire value of the `random_state` variable. It is not a very good practice because knowing the generator parameters and the last returned random value, you can easily calculate the entire sequence of next generated values.

shelib01 replied:
@piotr-now . I have read your and Alexander's explanations.Thank you :)
 Let's only add comment to explain

piotr-now replied:
No problem :). I have added an explanation about this 16-bit value.

piotr-now commented on include/mbedtls/platform_util.h:
```diff
@@ -263,8 +275,8 @@ uint32_t mbedtls_platform_random_in_range( size_t num );
- * \note        Currently the function is dependent of hardware providing an
- *              rng with MBEDTLS_ENTROPY_HARDWARE_ALT.
+ * \note        This function works only if the MBEDTLS_FI_COUNTERMEASURES flag
+ *              is defined in the configuration.
```
Added

piotr-now commented on library/platform_util.c:
```diff
@@ -172,37 +183,79 @@ int mbedtls_platform_memcmp( const void *buf1, const void *buf2, size_t num )
+/* This function implements a non-cryptographic random number generator based
+ * on the linear congruential generator algorithm. Additionally, if the
+ * MBEDTLS_ENTROPY_HARDWARE_ALT flag is defined, the seed is changed with using
+ * a hardware random number generator every RAND_SEED_LIFE number of iterations.
```
Added

piotr-now commented:
Force push after requested changes.

AndrzejKurek commented:
Please note that there are failures in the head job, which do not seem random.

shelib01 commented:
@piotr-now , did you look on Jenkins tests errors?

piotr-now commented:
> @piotr-now , did you look on Jenkins tests errors?

Yes, I'm working on fixing this

piotr-now commented:
I fixed the CI tests. The ones that remain are related to API change and Travis test bugs.

Code size changes:
|                 | Before | After | delta |
|-----------------|--------|-------|-------|
| libmbedcrypto.a | 19257  | 19465 | 208   |
| libmbedx509.a   | 4821   | 4821  | 0     |
| libmbedtls.a    | 16901  | 16901 | 0     |
| total           | 40979  | 41187 | 208   |

shelib01 commented:
@piotr-now , there are some comments to fix before merge:
1. https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-7d2c8f790aa5998c269656d95de4851cR239 , the function already can be eused without MBEDTLS_ENTROPY_HARDWARE_ALT flag
 2. https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-7d2c8f790aa5998c269656d95de4851cR257, the function already can be eused without hw_alt flagMBEDTLS_ENTROPY_HARDWARE_ALT
3. Comments to double random in https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-301de698d8af64dd1c923a3ab9fe1f07R232

piotr-now commented:
Force push after adding proposed comments and rebase to the baremetal.

piotr-now commented:
> @piotr-now , there are some comments to fix before merge:
> 
> 1. https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-7d2c8f790aa5998c269656d95de4851cR239 , the function already can be eused without MBEDTLS_ENTROPY_HARDWARE_ALT flag
> 2. https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-7d2c8f790aa5998c269656d95de4851cR257, the function already can be eused without hw_alt flagMBEDTLS_ENTROPY_HARDWARE_ALT
> 3. Comments to double random in https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-301de698d8af64dd1c923a3ab9fe1f07R232

@shelib01 I just added a comment about the last point here: https://github.com/ARMmbed/mbedtls/blob/585652772a8062f46a2d43ca39421ab8794a8b35/library/platform_util.c#L194
But I don't really understand the first two. Could I ask you for an explanation of them?

shelib01 commented:
> > @piotr-now , there are some comments to fix before merge:
> > 
> > 1. https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-7d2c8f790aa5998c269656d95de4851cR239 , the function already can be eused without MBEDTLS_ENTROPY_HARDWARE_ALT flag
> > 2. https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-7d2c8f790aa5998c269656d95de4851cR257, the function already can be eused without hw_alt flagMBEDTLS_ENTROPY_HARDWARE_ALT
> > 3. Comments to double random in https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-301de698d8af64dd1c923a3ab9fe1f07R232
> 
> @shelib01 I just added a comment about the last point here:
> 
> https://github.com/ARMmbed/mbedtls/blob/585652772a8062f46a2d43ca39421ab8794a8b35/library/platform_util.c#L194
> 
> 
> But I don't really understand the first two. Could I ask you for an explanation of them?

Sure :)
Lines 238 and and 257 in file platform_utils.h say that functions are dependent on MBEDTLS_ENTROPY_HARDWARE_ALT. I think that after random delay function changes the are not, and we should fix these comments

piotr-now commented:
> > > @piotr-now , there are some comments to fix before merge:
> > > 
> > > 1. https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-7d2c8f790aa5998c269656d95de4851cR239 , the function already can be eused without MBEDTLS_ENTROPY_HARDWARE_ALT flag
> > > 2. https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-7d2c8f790aa5998c269656d95de4851cR257, the function already can be eused without hw_alt flagMBEDTLS_ENTROPY_HARDWARE_ALT
> > > 3. Comments to double random in https://github.com/ARMmbed/mbedtls/pull/3537/files#diff-301de698d8af64dd1c923a3ab9fe1f07R232
> > 
> > 
> > @shelib01 I just added a comment about the last point here:
> > https://github.com/ARMmbed/mbedtls/blob/585652772a8062f46a2d43ca39421ab8794a8b35/library/platform_util.c#L194
> > 
> > But I don't really understand the first two. Could I ask you for an explanation of them?
> 
> Sure :)
> Lines 238 and and 257 in file platform_util.h say that functions are dependent on MBEDTLS_ENTROPY_HARDWARE_ALT. I think that after random delay function changes they are not, and we should fix these comments

Yes, good point. Thank you :). I also fixed a few other places in the `platform_utils.h` file

AndrzejKurek commented on library/platform_util.c:
Why +1?

piotr-now replied:
The `shift` value cannot be a zero, because some of tests with the ASanDbg options are failing on this line `rn_2 << ( 32 - shift )`. Shifting by 32 is treated as an error on the 32-bit architecture. Previously, during tests, there was never a situation where the value of `shift` was equal to zero so the tests were passed.

Increasing by one ensures that this value will always be different from zero, and at the same time the behavior of the entire function will not be significantly changed.
